File: matlab/plot_Pr.m
clear

trial = 28;
K = 20;
N = 1;
B = 0;
E = 1;

Pr_set = [0.01, 0.1, 0.3, 0.5, 1];


filename=['training_result/cmp_Pr_trial_' num2str(trial) '_K_' num2str(K) '_N_' num2str(N)  '_B_' num2str(B) '_E_' num2str(E) '.mat'];

load(filename);

filename=['training_result/cmp_Pr_trial_' num2str(trial) '_K_' num2str(K) '_N_' num2str(N)  '_B_' num2str(B) '_E_' num2str(E) '_NMSE.mat'];

load(filename);

linesize=1.5;
MarkerSize=8;
LineWidth=1.5;

figure

hold on
plot(Pr_set, test_accuracy(1, 1 : end), 'k--', 'LineWidth', LineWidth, 'MarkerSize', MarkerSize);
plot(Pr_set, test_accuracy(2, 1 : end), 'r-o', 'LineWidth', LineWidth, 'MarkerSize', MarkerSize, 'MarkerFaceColor', 'r');
plot(Pr_set, ones(length(Pr_set), 1) * mean(test_accuracy(3, 1: end)), '-^', 'Color', [0.4940 0.1840 0.5560], 'LineWidth', LineWidth, 'MarkerSize', MarkerSize, 'MarkerFaceColor', [0.4940 0.1840 0.5560]);
plot(Pr_set, test_accuracy(5, 1 : end), '-p', 'Color', [0.4660 0.6740 0.1880], 'LineWidth', LineWidth, 'MarkerSize', 2 + MarkerSize, 'MarkerFaceColor', [0.4660 0.6740 0.1880]);

set(get(gca, 'Children'), 'linewidth', 1.5)
set(gca, 'XTick', [0.01, 0.1, 0.3, 0.5, 1])
% xticklabels({'0.01', '0.1', '0.3', '0.5', '1'})
% set(gca, 'XLim', [Pr_set(1), Pr_set(end)])
set(gca, 'YTick', 0: 0.1: 0.9)
axis([0 Pr_set(end) 0 0.9])
axis([0.01 1 0.4 0.9])

grid on 
box on
hl = legend('Error-free channel', 'Proposed scheme', 'Conventional scheme', 'Existing scheme [26]');
set(hl,'Interpreter', 'latex', 'fontsize', 12, 'location', 'southeast')
xlabel('Maximum Relay Transmit Power $P_r$ (W)', 'Interpreter', 'latex', 'fontsize', 14);
ylabel('Test Accuracy','Interpreter', 'latex', 'fontsize', 14);

figure

hold on
plot(Pr_set, nmse(2, 1 : end), 'r-o', 'LineWidth', LineWidth, 'MarkerSize', MarkerSize, 'MarkerFaceColor', 'r');
plot(Pr_set, ones(length(Pr_set), 1) * mean(nmse(3, 1: end)), '-^', 'Color', [0.4940 0.1840 0.5560], 'LineWidth', LineWidth, 'MarkerSize', MarkerSize, 'MarkerFaceColor', [0.4940 0.1840 0.5560]);
plot(Pr_set, nmse(5, 1 : end), '-p', 'Color', [0.4660 0.6740 0.1880], 'LineWidth', LineWidth, 'MarkerSize', 2 + MarkerSize, 'MarkerFaceColor', [0.4660 0.6740 0.1880]);

set(get(gca, 'Children'), 'linewidth', 1.5)
set(gca, 'XTick', [0.01, 0.1, 0.3, 0.5, 1])
% xticklabels({'0.01', '0.1', '0.3', '0.5', '1'})
set(gca, 'XLim', [Pr_set(1), Pr_set(end)])
% set(gca, 'YTick', -10: 5: 10)
% axis([0 Pr_set(end) 0 0.9])
axis([0.01 1 -10 10])

grid on 
box on
hl = legend('Proposed scheme', 'Conventional scheme', 'Existing scheme [26]');
set(hl,'Interpreter', 'latex', 'fontsize', 12, 'location', 'southeast')
xlabel('Maximum Relay Transmit Power $P_r$ (W)', 'Interpreter', 'latex', 'fontsize', 14);
ylabel('Average NMSE (dB)','Interpreter', 'latex', 'fontsize', 14);

File: plot_Pr.py
# -*- coding: utf-8 -*-

import numpy as np
import matplotlib.pyplot as plt
import copy
import scipy.io as sio

if __name__ == '__main__':
    d = 21921
    trial = 50
    trial2 = 28
    K = 20
    N = 1
    SNR = 100
    B = 0
    E = 1
    lr = 0.05
    PL = 3.0
    P_r = 0.1
    iid = 1
    noniid_level = 2
    Pr_set = [0.01, 0.1, 0.3, 0.5, 1]

    test_accuracy = np.zeros([5, len(Pr_set)])
    training_loss = np.zeros([5, len(Pr_set)])
    nmse = np.zeros([5, len(Pr_set)])

    for i in range(len(Pr_set)):
        P_r = Pr_set[i]
        filename = 'store/trial_{}_K_{}_N_{}_B_{}_E_{}_lr_{}_SNR_{}_PL_{}_Pr_{}.npz'.format(trial, K, N, B, E, lr, SNR,
                                                                                            PL, P_r)

        a = np.load(filename, allow_pickle=True)
        res = a['arr_1']
        nmse1 = a['arr_3']
        nmse2 = a['arr_4']
        nmse4 = a['arr_6']

        nmse[1, i] = 10 * np.log10(np.mean(nmse1[~np.isnan(nmse1)]))
        nmse[2, i] = 10 * np.log10(np.mean(nmse2[~np.isnan(nmse2)]))
        nmse[4, i] = 10 * np.log10(np.mean(nmse4[~np.isnan(nmse4)]))

        res_CNN = {}
        for iter in range(trial2):
            if iter == 0:
                res_CNN = copy.deepcopy(res[0])
            else:
                for item in res_CNN.keys():
                    res_CNN[item] += copy.deepcopy(res[iter][item])

        for item in res_CNN.keys():
            res_CNN[item] = copy.deepcopy(res_CNN[item] / trial2)

        test_accuracy[0, i] = res_CNN['accuracy_test1'][1000]
        test_accuracy[1, i] = res_CNN['accuracy_test2'][500]
        test_accuracy[2, i] = res_CNN['accuracy_test3'][1000]
        test_accuracy[4, i] = res_CNN['accuracy_test5'][500]

    print(test_accuracy)
    print(training_loss)

    matfile = 'matlab/training_result/cmp_Pr_trial_{}_K_{}_N_{}_B_{}_E_{}.mat'.format(trial2, K, N, B, E)
    sio.savemat(matfile, mdict={'test_accuracy': test_accuracy})

    matfile2 = 'matlab/training_result/cmp_Pr_trial_{}_K_{}_N_{}_B_{}_E_{}_NMSE.mat'.format(trial2, K, N, B, E)
    sio.savemat(matfile2, mdict={'nmse': nmse})

    plt.plot(Pr_set, test_accuracy[0], 'r-')
    plt.plot(Pr_set, test_accuracy[1], 'b-')
    plt.plot(Pr_set, test_accuracy[2], 'g-')
    plt.plot(Pr_set, test_accuracy[4], 'y-')
    plt.legend(labels=['Error-Free', 'Proposed', 'Conventional', 'Existing Scheme'], loc='lower center',
               fontsize='x-large')
    plt.xlim([0.01, 1])
    plt.xticks(Pr_set)
    # plt.ylim([0.2, 0.9])
    # plt.yticks([0.5, 0.6, 0.7, 0.8, 0.9])
    plt.xlabel('Maximum relay transmit power $P_r$')
    plt.ylabel('Test Accuracy')
    plt.grid()

    plt.figure()

    plt.plot(Pr_set, nmse[1], 'b-')
    plt.plot(Pr_set, nmse[2], 'g-')
    plt.plot(Pr_set, nmse[4], 'y-')
    plt.legend(labels=['Proposed', 'Conventional', 'Existing Scheme'], loc='lower center',
               fontsize='x-large')
    plt.xlim([0.01, 1])
    plt.xticks(Pr_set)
    # plt.ylim([0.2, 0.9])
    # plt.yticks([0.5, 0.6, 0.7, 0.8, 0.9])
    plt.xlabel('Maximum relay transmit power $P_r$')
    plt.ylabel('Average MSE')

    plt.show()
